# Machine Learning With Spark ML
In this lab assignment, you will complete a project by going through the following steps:
1. Get the data.
2. Discover the data to gain insights.
3. Prepare the data for Machine Learning algorithms.
4. Select a model and train it.
5. Fine-tune your model.
6. Present your solution.

As a dataset, we use the California Housing Prices dataset from the StatLib repository. This dataset was based on data from the 1990 California census. The dataset has the following columns
1. `longitude`: a measure of how far west a house is (a higher value is farther west)
2. `latitude`: a measure of how far north a house is (a higher value is farther north)
3. `housing_,median_age`: median age of a house within a block (a lower number is a newer building)
4. `total_rooms`: total number of rooms within a block
5. `total_bedrooms`: total number of bedrooms within a block
6. `population`: total number of people residing within a block
7. `households`: total number of households, a group of people residing within a home unit, for a block
8. `median_income`: median income for households within a block of houses
9. `median_house_value`: median house value for households within a block
10. `ocean_proximity`: location of the house w.r.t ocean/sea

---
# 1. Get the data
Let's start the lab by loading the dataset. The can find the dataset at `data/housing.csv`. To infer column types automatically, when you are reading the file, you need to set `inferSchema` to true. Moreover enable the `header` option to read the columns' name from the file.

In [59]:
import findspark
findspark.init()
import pyspark
from pyspark.sql import SparkSession
spark = SparkSession.builder.getOrCreate()

In [60]:
# TODO: Replace <FILL IN> with appropriate code

housing = spark.read.format("csv").option("header", "true").option("inferSchema","true").load("data/housing.csv")

---
# 2. Discover the data to gain insights
Now it is time to take a look at the data. In this step we are going to take a look at the data a few different ways:
* See the schema and dimension of the dataset
* Look at the data itself
* Statistical summary of the attributes
* Breakdown of the data by the categorical attribute variable
* Find the correlation among different attributes
* Make new attributes by combining existing attributes

## 2.1. Schema and dimension
Print the schema of the dataset

In [61]:
# TODO: Replace <FILL IN> with appropriate code

housing.printSchema()

root
 |-- longitude: double (nullable = true)
 |-- latitude: double (nullable = true)
 |-- housing_median_age: double (nullable = true)
 |-- total_rooms: double (nullable = true)
 |-- total_bedrooms: double (nullable = true)
 |-- population: double (nullable = true)
 |-- households: double (nullable = true)
 |-- median_income: double (nullable = true)
 |-- median_house_value: double (nullable = true)
 |-- ocean_proximity: string (nullable = true)



Print the number of records in the dataset.

In [62]:
# TODO: Replace <FILL IN> with appropriate code

housing.count()

20640

## 2.2. Look at the data
Print the first five records of the dataset.

In [63]:
# TODO: Replace <FILL IN> with appropriate code

housing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|median_house_value|ocean_proximity|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|          452600.0|       NEAR BAY|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|          358500.0|       NEAR BAY|
|  -122.24|   37.85|              52.0|     1467.0|         190.0|     496.0|     177.0|       7.2574|          352100.0|       NEAR BAY|
|  -122.25|   37.85|              52.0|     1274.0|         235.0|     558.0|     219.0|       5.6431|          341300.0|       NEAR BAY|
|  -122.25|   37.85|              

Print the number of records with population more than 10000.

In [64]:
# TODO: Replace <FILL IN> with appropriate code
from pyspark.sql.functions import col
housing.filter(col("population")>10000).show()

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|median_house_value|ocean_proximity|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|  -121.92|   37.53|               7.0|    28258.0|        3864.0|   12203.0|    3701.0|       8.4045|          451100.0|      <1H OCEAN|
|  -117.78|   34.03|               8.0|    32054.0|        5290.0|   15507.0|    5050.0|       6.0191|          253900.0|      <1H OCEAN|
|  -117.87|   34.04|               7.0|    27700.0|        4179.0|   15037.0|    4072.0|       6.6288|          339700.0|      <1H OCEAN|
|  -117.88|   33.96|              16.0|    19059.0|        3079.0|   10988.0|    3061.0|       5.5469|          265200.0|      <1H OCEAN|
|  -118.78|   34.16|              

## 2.3. Statistical summary
Print a summary of the table statistics for the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. You can use the `describe` command.

In [65]:
# TODO: Replace <FILL IN> with appropriate code
housing.describe(["housing_median_age","total_rooms","median_house_value","population"]).show()

+-------+------------------+------------------+------------------+------------------+
|summary|housing_median_age|       total_rooms|median_house_value|        population|
+-------+------------------+------------------+------------------+------------------+
|  count|             20640|             20640|             20640|             20640|
|   mean|28.639486434108527|2635.7630813953488|206855.81690891474|1425.4767441860465|
| stddev| 12.58555761211163|2181.6152515827944|115395.61587441359|  1132.46212176534|
|    min|               1.0|               2.0|           14999.0|               3.0|
|    max|              52.0|           39320.0|          500001.0|           35682.0|
+-------+------------------+------------------+------------------+------------------+



Print the maximum age (`housing_median_age`), the minimum number of rooms (`total_rooms`), and the average of house values (`median_house_value`).

In [66]:
# TODO: Replace <FILL IN> with appropriate code

from pyspark.sql.functions import max, min, avg

housing.select(max("housing_median_age")).show()
housing.select(min("total_rooms")).show()
housing.select(avg("median_house_value")).show()

+-----------------------+
|max(housing_median_age)|
+-----------------------+
|                   52.0|
+-----------------------+

+----------------+
|min(total_rooms)|
+----------------+
|             2.0|
+----------------+

+-----------------------+
|avg(median_house_value)|
+-----------------------+
|     206855.81690891474|
+-----------------------+



## 2.4. Breakdown the data by categorical data
Print the number of houses in different areas (`ocean_proximity`), and sort them in descending order.

In [67]:
# TODO: Replace <FILL IN> with appropriate code

from pyspark.sql.functions import desc 
housing.groupBy("ocean_proximity").count().orderBy(desc("count")).show()

+---------------+-----+
|ocean_proximity|count|
+---------------+-----+
|      <1H OCEAN| 9136|
|         INLAND| 6551|
|     NEAR OCEAN| 2658|
|       NEAR BAY| 2290|
|         ISLAND|    5|
+---------------+-----+



Print the average value of the houses (`median_house_value`) in different areas (`ocean_proximity`), and call the new column `avg_value` when print it.

In [68]:
# TODO: Replace <FILL IN> with appropriate code

housing.groupBy("ocean_proximity").avg("median_house_value").withColumnRenamed("avg(median_house_value)","avg_value").show()

+---------------+------------------+
|ocean_proximity|         avg_value|
+---------------+------------------+
|         ISLAND|          380440.0|
|     NEAR OCEAN|249433.97742663656|
|       NEAR BAY|259212.31179039303|
|      <1H OCEAN|240084.28546409807|
|         INLAND|124805.39200122119|
+---------------+------------------+



Rewrite the above question in SQL.

In [69]:
# TODO: Replace <FILL IN> with appropriate code
housing.createOrReplaceTempView("df")
spark.sql("""SELECT ocean_proximity, avg(median_house_value) as avg_value FROM df GROUP BY ocean_proximity""").show()
# spark.sql(SELECT avg_value FROM df BY longtitude).show()

+---------------+------------------+
|ocean_proximity|         avg_value|
+---------------+------------------+
|         ISLAND|          380440.0|
|     NEAR OCEAN|249433.97742663656|
|       NEAR BAY|259212.31179039303|
|      <1H OCEAN|240084.28546409807|
|         INLAND|124805.39200122119|
+---------------+------------------+



In [70]:

#spark.sql("SELECT 1 + 1").show()

## 2.5. Correlation among attributes
Print the correlation among the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. To do so, first you need to put these attributes into one vector. Then, compute the standard correlation coefficient (Pearson) between every pair of attributes in this new vector. To make a vector of these attributes, you can use the `VectorAssembler` Transformer.

In [71]:
# TODO: Replace <FILL IN> with appropriate code

from pyspark.ml.feature import VectorAssembler

va = VectorAssembler().setInputCols(["housing_median_age","total_rooms","median_house_value","population"]).setOutputCol("feature")

housingAttrs = va.transform(housing)

housingAttrs.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+--------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|median_house_value|ocean_proximity|             feature|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+--------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|          452600.0|       NEAR BAY|[41.0,880.0,45260...|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|          358500.0|       NEAR BAY|[21.0,7099.0,3585...|
|  -122.24|   37.85|              52.0|     1467.0|         190.0|     496.0|     177.0|       7.2574|          352100.0|       NEAR BAY|[52.0,1467.0,3521...|
|  -122.25|   37.85|              52.0|     12

In [72]:
# TODO: Replace <FILL IN> with appropriate code
from pyspark.ml.stat import Correlation
#import org.apache.spark.ml.linalg.Matrix
#import org.apache.spark.ml.stat.Correlation
#import org.apache.spark.sql.Row
#housingAttrs.select(corr("feature","housing_median_age")).show()
#Row(coeff: Matrix) = Correlation.corr(housingAttrs)#create rows
#print(s"The standard correlation coefficient:\n ${coeff}")
corr_matrix = Correlation.corr(housingAttrs,"feature")
corr_matrix.collect()[0]["pearson({})".format("feature")].values

array([ 1.        , -0.3612622 ,  0.10562341, -0.29624424, -0.3612622 ,
        1.        ,  0.13415311,  0.85712597,  0.10562341,  0.13415311,
        1.        , -0.02464968, -0.29624424,  0.85712597, -0.02464968,
        1.        ])

## 2.6. Combine and make new attributes
Now, let's try out various attribute combinations. In the given dataset, the total number of rooms in a block is not very useful, if we don't know how many households there are. What we really want is the number of rooms per household. Similarly, the total number of bedrooms by itself is not very useful, and we want to compare it to the number of rooms. And the population per household seems like also an interesting attribute combination to look at. To do so, add the three new columns to the dataset as below. We will call the new dataset the `housingExtra`.
```
rooms_per_household = total_rooms / households
bedrooms_per_room = total_bedrooms / total_rooms
population_per_household = population / households
```

In [73]:
# TODO: Replace <FILL IN> with appropriate code

housingCol1 = housing.withColumn("rooms_per_household", housing.total_rooms / housing.households)
housingCol2 = housingCol1.withColumn("bedrooms_per_room", housing.total_bedrooms / housing.total_rooms)
housingExtra = housingCol2.withColumn("population_per_household", housing.population / housing.households)

housingExtra.select("rooms_per_household", "bedrooms_per_room", "population_per_household").show(5)

+-------------------+-------------------+------------------------+
|rooms_per_household|  bedrooms_per_room|population_per_household|
+-------------------+-------------------+------------------------+
|  6.984126984126984|0.14659090909090908|      2.5555555555555554|
|  6.238137082601054|0.15579659106916466|       2.109841827768014|
|  8.288135593220339|0.12951601908657123|      2.8022598870056497|
| 5.8173515981735155|0.18445839874411302|       2.547945205479452|
|  6.281853281853282| 0.1720958819913952|      2.1814671814671813|
+-------------------+-------------------+------------------------+
only showing top 5 rows



---
## 3. Prepare the data for Machine Learning algorithms
Before going through the Machine Learning steps, let's first rename the label column from `median_house_value` to `label`.

In [74]:
# TODO: Replace <FILL IN> with appropriate code

renamedHousing = housingExtra.withColumnRenamed("median_house_value","label")
renamedHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|358500.0|       NEAR BAY|  6.238137082601054|0.15579659106916466|       2.109841827768014|
|  -122.24|   37.85|              5

Now, we want to separate the numerical attributes from the categorical attribute (`ocean_proximity`) and keep their column names in two different lists. Moreover, sice we don't want to apply the same transformations to the predictors (features) and the label, we should remove the label attribute from the list of predictors. 

In [75]:
# label columns
colLabel = "label"

# categorical columns
colCat = "ocean_proximity"

# numerical columns
colNum = renamedHousing.drop(colLabel).drop(colCat)
colNum.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+-------------------+-------------------+------------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|rooms_per_household|  bedrooms_per_room|population_per_household|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+-------------------+-------------------+------------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|  6.984126984126984|0.14659090909090908|      2.5555555555555554|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|  6.238137082601054|0.15579659106916466|       2.109841827768014|
|  -122.24|   37.85|              52.0|     1467.0|         190.0|     496.0|     177.0|       7.2574|  8.288135593220339|0.12951601908657123|      2.8022598870

## 3.1. Prepare continuse attributes
### Data cleaning
Most Machine Learning algorithms cannot work with missing features, so we should take care of them. As a first step, let's find the columns with missing values in the numerical attributes. To do so, we can print the number of missing values of each continues attributes, listed in `colNum`.

In [76]:
# TODO: Replace <FILL IN> with appropriate code

# for (c <- colNum) {
#     <FILL IN>
# }
for c in colNum :
    print(c, colNum.filter(c.isNull()).count())

Column<'longitude'> 0
Column<'latitude'> 0
Column<'housing_median_age'> 0
Column<'total_rooms'> 0
Column<'total_bedrooms'> 207
Column<'population'> 0
Column<'households'> 0
Column<'median_income'> 0
Column<'rooms_per_household'> 0
Column<'bedrooms_per_room'> 207
Column<'population_per_household'> 0


As we observerd above, the `total_bedrooms` and `bedrooms_per_room` attributes have some missing values. One way to take care of missing values is to use the `Imputer` Transformer, which completes missing values in a dataset, either using the mean or the median of the columns in which the missing values are located. To use it, you need to create an `Imputer` instance, specifying that you want to replace each attribute's missing values with the "median" of that attribute.

In [77]:
# TODO: Replace <FILL IN> with appropriate code

#import org.apache.spark.ml.feature.Imputer
from pyspark.ml.feature import Imputer
imputer = Imputer().setStrategy("median").setInputCols(["total_bedrooms","bedrooms_per_room"]).setOutputCols(["total_bedrooms","bedrooms_per_room"])                
imputedHousing = imputer.fit(renamedHousing).transform(renamedHousing)

imputedHousing.select("total_bedrooms", "bedrooms_per_room").show(5)

+--------------+-------------------+
|total_bedrooms|  bedrooms_per_room|
+--------------+-------------------+
|         129.0|0.14659090909090908|
|        1106.0|0.15579659106916466|
|         190.0|0.12951601908657123|
|         235.0|0.18445839874411302|
|         280.0| 0.1720958819913952|
+--------------+-------------------+
only showing top 5 rows



### Scaling
One of the most important transformations you need to apply to your data is feature scaling. With few exceptions, Machine Learning algorithms don't perform well when the input numerical attributes have very different scales. This is the case for the housing data: the total number of rooms ranges from about 6 to 39,320, while the median incomes only range from 0 to 15. Note that scaling the label attribues is generally not required.

One way to get all attributes to have the same scale is to use standardization. In standardization, for each value, first it subtracts the mean value (so standardized values always have a zero mean), and then it divides by the variance so that the resulting distribution has unit variance. To do this, we can use the `StandardScaler` Estimator. To use `StandardScaler`, again we need to convert all the numerical attributes into a big vectore of features using `VectorAssembler`, and then call `StandardScaler` on that vactor.

In [78]:
# TODO: Replace <FILL IN> with appropriate code

#import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}
from pyspark.ml.feature import StandardScaler

va = VectorAssembler().setInputCols(colNum.columns).setOutputCol("imputed_feature")
featuredHousing = va.transform(imputedHousing)

scaler = StandardScaler().setInputCol("imputed_feature").setOutputCol("scaled_feature")
scaledHousing = scaler.fit(featuredHousing).transform(featuredHousing)

scaledHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|     imputed_feature|      scaled_feature|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|[-122.23,37.88,41...|[-61.007269596069...|
|  -122.22|   37.86|              21.0|     7099.0|        1

## 3.2. Prepare categorical attributes
After imputing and scaling the continuse attributes, we should take care of the categorical attributes. Let's first print the number of distict values of the categirical attribute `ocean_proximity`.

In [79]:
#TODO: Replace <FILL IN> with appropriate code
from pyspark.sql.functions import count_distinct
renamedHousing.select(count_distinct("ocean_proximity")).show()

+-------------------------------+
|count(DISTINCT ocean_proximity)|
+-------------------------------+
|                              5|
+-------------------------------+



### String indexer
Most Machine Learning algorithms prefer to work with numbers. So let's convert the categorical attribute `ocean_proximity` to numbers. To do so, we can use the `StringIndexer` that encodes a string column of labels to a column of label indices. The indices are in [0, numLabels), ordered by label frequencies, so the most frequent label gets index 0.

In [80]:
# TODO: Replace <FILL IN> with appropriate code

# import org.apache.spark.ml.feature.StringIndexer
from pyspark.ml.feature import StringIndexer
indexer = StringIndexer().setInputCol("ocean_proximity").setOutputCol("ocean_Int")
idxHousing = indexer.fit(renamedHousing).transform(renamedHousing)

idxHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+---------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|ocean_Int|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+---------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|      3.0|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|358500.0|       NEAR BAY|  6.238137082601054|0.15579659106916466|       2.109841827768

Now we can use this numerical data in any Machine Learning algorithm. You can look at the mapping that this encoder has learned using the `labels` method: "<1H OCEAN" is mapped to 0, "INLAND" is mapped to 1, etc.

In [81]:
indexer.fit(renamedHousing).labels

['<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'NEAR BAY', 'ISLAND']

### One-hot encoding
Now, convert the label indices built in the last step into one-hot vectors. To do this, you can take advantage of the `OneHotEncoderEstimator` Estimator.

In [82]:
# TODO: Replace <FILL IN> with appropriate code

#import org.apache.spark.ml.feature.OneHotEncoderEstimator
from pyspark.ml.feature import OneHotEncoder
encoder = OneHotEncoder().setInputCol("ocean_Int").setOutputCol("ocean_one-hot")
ohHousing = encoder.fit(idxHousing).transform(idxHousing)
# val encoder = new OneHotEncoderEstimator().<FILL IN>
# val ohHousing = encoder.fit(idxHousing).transform(idxHousing)

ohHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+---------+-------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|ocean_Int|ocean_one-hot|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+---------+-------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|      3.0|(4,[3],[1.0])|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|358500.0|       NEAR BAY|  6.2

---
# 4. Pipeline
As you can see, there are many data transformation steps that need to be executed in the right order. For example, you called the `Imputer`, `VectorAssembler`, and `StandardScaler` from left to right. However, we can use the `Pipeline` class to define a sequence of Transformers/Estimators, and run them in order. A `Pipeline` is an `Estimator`, thus, after a Pipeline's `fit()` method runs, it produces a `PipelineModel`, which is a `Transformer`.

Now, let's create a pipeline called `numPipeline` to call the numerical transformers you built above (`imputer`, `va`, and `scaler`) in the right order from left to right, as well as a pipeline called `catPipeline` to call the categorical transformers (`indexer` and `encoder`). Then, put these two pipelines `numPipeline` and `catPipeline` into one pipeline.

In [83]:
# TODO: Replace <FILL IN> with appropriate code

from pyspark.ml import Pipeline, PipelineModel
# import org.apache.spark.ml.{Pipeline, PipelineModel, PipelineStage}
numPipeline = Pipeline().setStages([imputer, va, scaler])
catPipeline = Pipeline().setStages([indexer, encoder])
pipeline = Pipeline().setStages([numPipeline, catPipeline])
newHousing = pipeline.fit(renamedHousing).transform(renamedHousing)
# val numPipeline = <FILL IN>
# val catPipeline = <FILL IN>
# val pipeline = new Pipeline().setStages(Array(numPipeline, catPipeline))
# val newHousing = pipeline.fit(renamedHousing).transform(renamedHousing)

newHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+---------+-------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|     imputed_feature|      scaled_feature|ocean_Int|ocean_one-hot|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+---------+-------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|[-122.23,37.88,41...|[-61.00726

Now, use `VectorAssembler` to put all attributes of the final dataset `newHousing` into a big vector, and call the new column `features`.

In [84]:
# TODO: Replace <FILL IN> with appropriate code

va2 = VectorAssembler().setInputCols(["scaled_feature", "ocean_one-hot"]).setOutputCol("features")
dataset = va2.transform(newHousing).select("features", "label")

dataset.show(5)

+--------------------+--------+
|            features|   label|
+--------------------+--------+
|[-61.007269596069...|452600.0|
|[-61.002278409814...|358500.0|
|[-61.012260782324...|352100.0|
|[-61.017251968579...|341300.0|
|[-61.017251968579...|342200.0|
+--------------------+--------+
only showing top 5 rows



---
# 5. Make a model
Here we going to make four different regression models:
* Linear regression model
* Decission tree regression
* Random forest regression
* Gradient-booster forest regression

But, before giving the data to train a Machine Learning model, let's first split the data into training dataset (`trainSet`) with 80% of the whole data, and test dataset (`testSet`) with 20% of it.

In [85]:
# TODO: Replace <FILL IN> with appropriate code

[trainSet, testSet] = dataset.randomSplit([0.8, 0.2])

## 5.1. Linear regression model
Now, train a Linear Regression model using the `LinearRegression` class. Then, print the coefficients and intercept of the model, as well as the summary of the model over the training set by calling the `summary` method.

In [86]:
# // TODO: Replace <FILL IN> with appropriate code

# import org.apache.spark.ml.regression.LinearRegression

# // train the model
# val lr = <FILL IN>
# val lrModel = lr.<FILL IN>
# val trainingSummary = lrModel.summary

# println(s"Coefficients: <FILL IN>, Intercept: <FILL IN>")
# println(s"RMSE: <FILL IN>")
from pyspark.ml.regression import LinearRegression
# train the model
lr = LinearRegression()
lrModel = lr.fit(trainSet)

In [87]:
print(lrModel.coefficients)
print(lrModel.intercept)
print(lrModel.summary.rootMeanSquaredError)

[-56764.53108402594,-57328.8774213202,12780.5040231277,6765.82391713553,3201.059796969359,-44495.08911762411,39475.62135098883,78556.22878315115,6969.68126589621,16982.952266064513,-340.23752146214446,-166725.86085032913,-200237.20055643705,-162606.79017120402,-171261.6133139131]
-2322396.7823316776
67625.98296097975


Now, use `RegressionEvaluator` to measure the root-mean-square-erroe (RMSE) of the model on the test dataset.

In [88]:
# TODO: Replace <FILL IN> with appropriate code

# import org.apache.spark.ml.evaluation.RegressionEvaluator

# // make predictions on the test data
# val predictions = lrModel.<FILL IN>
# predictions.select("prediction", "label", "features").show(5)

# // select (prediction, true label) and compute test error.
# val evaluator = new RegressionEvaluator().<FILL IN>
# val rmse = evaluator.<FILL IN>
# println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

from pyspark.ml.evaluation import RegressionEvaluator
predictions = lrModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

evaluator = RegressionEvaluator()
rmse = evaluator.evaluate(predictions)
print(rmse)

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|103253.89100639522| 85800.0|[-62.040445150874...|
|141705.29281402146|103600.0|[-62.040445150874...|
|106442.68248020625| 66900.0|[-61.995524474579...|
|162855.67115859129| 82800.0|[-61.975559729558...|
|151521.93609251268| 74100.0|[-61.970568543303...|
+------------------+--------+--------------------+
only showing top 5 rows

68979.7903682375


## 5.2. Decision tree classification
Repeat what you have done on Regression Model to build a Decision Tree model. Use the `DecisionTreeRegressor` to make a model and then measure its RMSE on the test dataset.

In [89]:
# // TODO: Replace <FILL IN> with appropriate code

# import org.apache.spark.ml.regression.DecisionTreeRegressor
# import org.apache.spark.ml.evaluation.RegressionEvaluator

# val dt = new DecisionTreeRegressor().<FILL IN>

# // train the model
# val dtModel = dt.<FILL IN>

# // make predictions on the test data
# val predictions = dtModel.<FILL IN>
# predictions.select("prediction", "label", "features").show(5)

# // select (prediction, true label) and compute test error
# val evaluator = new RegressionEvaluator().<FILL IN>
# val rmse = evaluator.<FILL IN>
# println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

from pyspark.ml.regression import DecisionTreeRegressor
dt = DecisionTreeRegressor().setFeaturesCol("features").setLabelCol("label")
dtMoedl = dt.fit(trainSet)

predictions = dtMoedl.transform(testSet)
predictions.select("prediction", "label", "features").show(5)
rmse = evaluator.evaluate(predictions)
print(rmse)

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|138149.89894736843| 85800.0|[-62.040445150874...|
| 166936.5281848929|103600.0|[-62.040445150874...|
|138149.89894736843| 66900.0|[-61.995524474579...|
| 166936.5281848929| 82800.0|[-61.975559729558...|
|149143.31797235023| 74100.0|[-61.970568543303...|
+------------------+--------+--------------------+
only showing top 5 rows

67210.2425772272


## 5.3. Random forest regression
Let's try the test error on a Random Forest Model. Youcan use the `RandomForestRegressor` to make a Random Forest model.

In [90]:
# // TODO: Replace <FILL IN> with appropriate code

# import org.apache.spark.ml.regression.RandomForestRegressor
# import org.apache.spark.ml.evaluation.RegressionEvaluator

# val rf = new RandomForestRegressor().<FILL IN>

# // train the model
# val rfModel = rf.<FILL IN>

# // make predictions on the test data
# val predictions = rfModel.<FILL IN>
# predictions.select("prediction", "label", "features").show(5)

# // select (prediction, true label) and compute test error
# val evaluator = new RegressionEvaluator().<FILL IN>
# val rmse = evaluator.<FILL IN>
# println(s"Root Mean Squared Error (RMSE) on test data = $rmse")
from pyspark.ml.regression import RandomForestRegressor
rf = RandomForestRegressor().setFeaturesCol("features").setLabelCol("label")

rfModel = rf.fit(trainSet)
predictions = rfModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)
rmse = evaluator.evaluate(predictions)
print(rmse)

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|151603.29115478983| 85800.0|[-62.040445150874...|
|168935.66432622526|103600.0|[-62.040445150874...|
| 153076.9988363468| 66900.0|[-61.995524474579...|
| 162076.8884287946| 82800.0|[-61.975559729558...|
|169196.55847037578| 74100.0|[-61.970568543303...|
+------------------+--------+--------------------+
only showing top 5 rows

65588.64185573056


## 5.4. Gradient-boosted tree regression
Fianlly, we want to build a Gradient-boosted Tree Regression model and test the RMSE of the test data. Use the `GBTRegressor` to build the model.

In [91]:
# // TODO: Replace <FILL IN> with appropriate code

# import org.apache.spark.ml.regression.GBTRegressor
# import org.apache.spark.ml.evaluation.RegressionEvaluator

# val gb = new GBTRegressor().<FILL IN>

# // train the model
# val gbModel = gb.<FILL IN>

# // make predictions on the test data
# val predictions = gbModel.<FILL IN>
# predictions.select("prediction", "label", "features").show(5)

# // select (prediction, true label) and compute test error
# val evaluator = new RegressionEvaluator().<FILL IN>
# val rmse = evaluator.<FILL IN>
# println(s"Root Mean Squared Error (RMSE) on test data = $rmse")
from pyspark.ml.regression import GBTRegressor
gb = GBTRegressor().setFeaturesCol("features").setLabelCol("label")

gbModel = gb.fit(trainSet)

predictions = gbModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

rmse = evaluator.evaluate(predictions)

print(rmse)

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
| 94235.77625292915| 85800.0|[-62.040445150874...|
|122919.78816113572|103600.0|[-62.040445150874...|
| 99141.36868858915| 66900.0|[-61.995524474579...|
|112839.52124000133| 82800.0|[-61.975559729558...|
| 97482.35978475207| 74100.0|[-61.970568543303...|
+------------------+--------+--------------------+
only showing top 5 rows

54933.99887106847


---
# 6. Hyperparameter tuning
An important task in Machie Learning is model selection, or using data to find the best model or parameters for a given task. This is also called tuning. Tuning may be done for individual Estimators such as LinearRegression, or for entire Pipelines which include multiple algorithms, featurization, and other steps. Users can tune an entire Pipeline at once, rather than tuning each element in the Pipeline separately. MLlib supports model selection tools, such as `CrossValidator`. These tools require the following items:
* Estimator: algorithm or Pipeline to tune (`setEstimator`)
* Set of ParamMaps: parameters to choose from, sometimes called a "parameter grid" to search over (`setEstimatorParamMaps`)
* Evaluator: metric to measure how well a fitted Model does on held-out test data (`setEvaluator`)

`CrossValidator` begins by splitting the dataset into a set of folds, which are used as separate training and test datasets. For example with `k=3` folds, `CrossValidator` will generate 3 (training, test) dataset pairs, each of which uses 2/3 of the data for training and 1/3 for testing. To evaluate a particular `ParamMap`, `CrossValidator` computes the average evaluation metric for the 3 Models produced by fitting the Estimator on the 3 different (training, test) dataset pairs. After identifying the best `ParamMap`, `CrossValidator` finally re-fits the Estimator using the best ParamMap and the entire dataset.

Below, use the `CrossValidator` to select the best Random Forest model. To do so, you need to define a grid of parameters. Let's say we want to do the search among the different number of trees (1, 5, and 10), and different tree depth (5, 10, and 15).

In [92]:
# // TODO: Replace <FILL IN> with appropriate code

# import org.apache.spark.ml.tuning.ParamGridBuilder
# import org.apache.spark.ml.evaluation.RegressionEvaluator
# import org.apache.spark.ml.tuning.CrossValidator

# val paramGrid = new ParamGridBuilder().<FILL IN>

# val evaluator = new RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName(<FILL IN>)
# val cv = new CrossValidator().<FILL IN>
# val cvModel = cv.<FILL IN>

# val predictions = cvModel.<FILL IN>
# predictions.select("prediction", "label", "features").show(5)

# val rmse = evaluator.<FILL IN>
# println(s"Root Mean Squared Error (RMSE) on test data = $rmse")
from pyspark.ml.tuning import ParamGridBuilder, CrossValidator

paramGrid = ParamGridBuilder().addGrid(rf.numTrees, [1,5,10]).addGrid(rf.maxDepth,[5,10,15]).build()
evaluator = RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("rmse")
cv = CrossValidator().setEstimator(rf).setEstimatorParamMaps(paramGrid).setEvaluator(evaluator)

cvModel = cv.fit(trainSet)

predictions = cvModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

rmse = evaluator.evaluate(predictions)

print(rmse)

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
| 87290.32608695651| 85800.0|[-62.040445150874...|
|          100710.0|103600.0|[-62.040445150874...|
| 89721.66666666667| 66900.0|[-61.995524474579...|
| 85468.30128205128| 82800.0|[-61.975559729558...|
|103555.83333333333| 74100.0|[-61.970568543303...|
+------------------+--------+--------------------+
only showing top 5 rows

51761.89793786124


---
# 7. An End-to-End Classification Test
As the last step, you are given a dataset called `data/ccdefault.csv`. The dataset represents default of credit card clients. It has 30,000 cases and 24 different attributes. More details about the dataset is available at `data/ccdefault.txt`. In this task you should make three models, compare their results and conclude the ideal solution. Here are the suggested steps:
1. Load the data.
2. Carry out some exploratory analyses (e.g., how various features and the target variable are distributed).
3. Train a model to predict the target variable (risk of `default`).
  - Employ three different models (logistic regression, decision tree, and random forest).
  - Compare the models' performances (e.g., AUC).
  - Defend your choice of best model (e.g., what are the strength and weaknesses of each of these models?).
4. What more would you do with this data? Anything to help you devise a better solution?

This research aimed at the case of customers' default payments in Taiwan and compares the predictive accuracy of probability of default among six data mining methods. From the perspective of risk management, the result of predictive accuracy of the estimated probability of default will be more valuable than the binary result of classification - credible or not credible clients. Because the real probability of default is unknown, this study presented the novel "Sorting Smoothing Method" to estimate the real probability of default. With the real probability of default as the response variable (Y), and the predictive probability of default as the independent variable (X), the simple linear regression result (Y = A + BX) shows that the forecasting model produced by artificial neural network has the highest coefficient of determination; its regression intercept (A) is close to zero, and regression coefficient (B) to one. Therefore, among the six data mining techniques, artificial neural network is the only one that can accurately estimate the real probability of default.

---
# 7.1. Get the data
Let's start the lab by loading the dataset. The can find the dataset at `data/ccdefault.csv`. To infer column types automatically, when you are reading the file, you need to set `inferSchema` to true. Moreover enable the `header` option to read the columns' name from the file.

In [93]:
ccdefault = spark.read.format("csv").option("header",True).option("inferSchema",True).load("data/ccdefault.csv")

# 7.2 Discover the data to gain insights


## 7.2.1 Schema and dimension

In [94]:
ccdefault.printSchema()

root
 |-- ID: integer (nullable = true)
 |-- LIMIT_BAL: integer (nullable = true)
 |-- SEX: integer (nullable = true)
 |-- EDUCATION: integer (nullable = true)
 |-- MARRIAGE: integer (nullable = true)
 |-- AGE: integer (nullable = true)
 |-- PAY_0: integer (nullable = true)
 |-- PAY_2: integer (nullable = true)
 |-- PAY_3: integer (nullable = true)
 |-- PAY_4: integer (nullable = true)
 |-- PAY_5: integer (nullable = true)
 |-- PAY_6: integer (nullable = true)
 |-- BILL_AMT1: integer (nullable = true)
 |-- BILL_AMT2: integer (nullable = true)
 |-- BILL_AMT3: integer (nullable = true)
 |-- BILL_AMT4: integer (nullable = true)
 |-- BILL_AMT5: integer (nullable = true)
 |-- BILL_AMT6: integer (nullable = true)
 |-- PAY_AMT1: integer (nullable = true)
 |-- PAY_AMT2: integer (nullable = true)
 |-- PAY_AMT3: integer (nullable = true)
 |-- PAY_AMT4: integer (nullable = true)
 |-- PAY_AMT5: integer (nullable = true)
 |-- PAY_AMT6: integer (nullable = true)
 |-- DEFAULT: integer (nullable = tru

## 7.2.2 Look at the data

In [95]:
ccdefault.show(5)

+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+
| ID|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|DEFAULT|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+
|  1|    20000|  2|        2|       1| 24|    2|    2|   -1|   -1|   -2|   -2|     3913|     3102|      689|        0|        0|        0|       0|     689|       0|       0|       0|       0|      1|
|  2|   120000|  2|        2|       2| 26|   -1|    2|    0|    0|    0|    2|     2682|     1725|     2682|     3272|     3455|     3261|       0|    1000|    1000|    1000|       0|    2000|    

## 7.2.3 Correlation among attributes

In [96]:
from pyspark.ml.feature import VectorAssembler

va = VectorAssembler().setInputCols(["LIMIT_BAL","SEX","EDUCATION","MARRIAGE","AGE"]).setOutputCol("feature")

ccAttrs = va.transform(ccdefault)

ccAttrs.show(5)

from pyspark.ml.stat import Correlation
corr_matrix = Correlation.corr(ccAttrs,"feature")
corr_matrix.collect()[0]["pearson({})".format("feature")].values

+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+--------------------+
| ID|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|DEFAULT|             feature|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+--------------------+
|  1|    20000|  2|        2|       1| 24|    2|    2|   -1|   -1|   -2|   -2|     3913|     3102|      689|        0|        0|        0|       0|     689|       0|       0|       0|       0|      1|[20000.0,2.0,2.0,...|
|  2|   120000|  2|        2|       2| 26|   -1|    2|    0|    0|    0|    2|     2682|     1725|     2682|    

array([ 1.        ,  0.02475524, -0.2191607 , -0.10813941,  0.1447128 ,
        0.02475524,  1.        ,  0.01423194, -0.03138884, -0.09087365,
       -0.2191607 ,  0.01423194,  1.        , -0.14346434,  0.17506066,
       -0.10813941, -0.03138884, -0.14346434,  1.        , -0.41416992,
        0.1447128 , -0.09087365,  0.17506066, -0.41416992,  1.        ])

## 7.2.4. Breakdown the data by categorical data

In [97]:
from pyspark.sql.functions import desc 
ccdefault.groupBy("SEX").avg("LIMIT_BAL").withColumnRenamed("avg(LIMIT_BAL)", "LIMIT_BAL_avg").sort(desc("LIMIT_BAL_avg")).show()

+---+------------------+
|SEX|     LIMIT_BAL_avg|
+---+------------------+
|  2|170086.46201413427|
|  1| 163519.8250336474|
+---+------------------+



## 7.2.5 Exploratory analyses

In [98]:
# Limit_BAL
ccdefault.select("LIMIT_BAL").describe().show()

+-------+------------------+
|summary|         LIMIT_BAL|
+-------+------------------+
|  count|             30000|
|   mean|167484.32266666667|
| stddev|129747.66156720246|
|    min|             10000|
|    max|           1000000|
+-------+------------------+



In [99]:
# SEX
sex_male = ccdefault.where("SEX = 1").count()
sex_female = ccdefault.where("SEX = 2").count()
print(f'male: {sex_male}, female: {sex_female}')

male: 11888, female: 18112


In [100]:
# EDUCATION
education_graduate_school = ccdefault.where("EDUCATION = 1").count()
education_uniersity = ccdefault.where("EDUCATION = 2").count()
education_high_school = ccdefault.where("EDUCATION = 3").count()
education_others = ccdefault.where("EDUCATION = 4").count()
print(f'graduate school: {education_graduate_school}, university: {education_uniersity}, high school: {education_high_school}, others: {education_others}')

graduate school: 10585, university: 14030, high school: 4917, others: 123


In [101]:
# MARRIAGE
marriage_married = ccdefault.where("MARRIAGE = 1").count()
marriage_single = ccdefault.where("MARRIAGE = 2").count()
marriage_other = ccdefault.where("MARRIAGE = 3").count()
print(f'married: {marriage_married}, single: {marriage_single}, other: {marriage_other}')

married: 13659, single: 15964, other: 323


In [102]:
# AGE
ccdefault.select("AGE").describe().show()

+-------+-----------------+
|summary|              AGE|
+-------+-----------------+
|  count|            30000|
|   mean|          35.4855|
| stddev|9.217904068090155|
|    min|               21|
|    max|               79|
+-------+-----------------+



In [103]:
# DEFAULT
default = ccdefault.where("DEFAULT = 1").count()
print(f'default: {default}')

default: 6636


# 7.3 Prepare the data for Machine Learning algorithms

We need to separate the column "DEFAULT" and "ID"

In [104]:
cc_data = ccdefault.drop("ID")
cc_feature_origin = ccdefault.drop("DEFAULT", "ID")
cc_feature_origin.show(5)

+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+
|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|
+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+
|    20000|  2|        2|       1| 24|    2|    2|   -1|   -1|   -2|   -2|     3913|     3102|      689|        0|        0|        0|       0|     689|       0|       0|       0|       0|
|   120000|  2|        2|       2| 26|   -1|    2|    0|    0|    0|    2|     2682|     1725|     2682|     3272|     3455|     3261|       0|    1000|    1000|    1000|       0|    2000|
|    90000|  2|        2|       2| 34|    0|    0|    0

## 7.3.1 Prepare continuse attributes
## Data cleaning

In [105]:
for cc in cc_feature_origin :
    print(cc, cc_feature_origin.filter(cc.isNull()).count()) 

Column<'LIMIT_BAL'> 0
Column<'SEX'> 0
Column<'EDUCATION'> 0
Column<'MARRIAGE'> 0
Column<'AGE'> 0
Column<'PAY_0'> 0
Column<'PAY_2'> 0
Column<'PAY_3'> 0
Column<'PAY_4'> 0
Column<'PAY_5'> 0
Column<'PAY_6'> 0
Column<'BILL_AMT1'> 0
Column<'BILL_AMT2'> 0
Column<'BILL_AMT3'> 0
Column<'BILL_AMT4'> 0
Column<'BILL_AMT5'> 0
Column<'BILL_AMT6'> 0
Column<'PAY_AMT1'> 0
Column<'PAY_AMT2'> 0
Column<'PAY_AMT3'> 0
Column<'PAY_AMT4'> 0
Column<'PAY_AMT5'> 0
Column<'PAY_AMT6'> 0


As the dataset does not has any missing values, we do not need to impute data.

In [106]:
## Scaling

In [107]:
from pyspark.ml.feature import StandardScaler
cc_va = VectorAssembler().setInputCols(cc_feature_origin.columns).setOutputCol("feature_origin")
featured_cc = cc_va.transform(cc_feature_origin)

scaler_cc = StandardScaler().setInputCol("feature_origin").setOutputCol("featured_scaled")
scaled_cc = scaler_cc.fit(featured_cc).transform(featured_cc)
scaled_cc.show(5)

+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+--------------------+--------------------+
|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|      feature_origin|     featured_scaled|
+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+--------------------+--------------------+
|    20000|  2|        2|       1| 24|    2|    2|   -1|   -1|   -2|   -2|     3913|     3102|      689|        0|        0|        0|       0|     689|       0|       0|       0|       0|[20000.0,2.0,2.0,...|[0.15414535998894...|
|   120000|  2|        2|       2| 26|   -1|    2|    0|    0|    0|    2|  

## One-hot encoding

We need to convert the feature "SEX", "EDUCATION", "MARRIAGE" into one-hot vectors.

In [108]:
from pyspark.ml.feature import OneHotEncoder
encoder_sex = OneHotEncoder().setInputCol("SEX").setOutputCol("SEX_one-hot")
encoder_education = OneHotEncoder().setInputCol("EDUCATION").setOutputCol("EDUCATION_one-hot")
encoder_marriage = OneHotEncoder().setInputCol("MARRIAGE").setOutputCol("MARRIAGE_one-hot")
ohSEX = encoder_sex.fit(cc_feature_origin).transform(cc_feature_origin)
ohEDUCATION = encoder_education.fit(cc_feature_origin).transform(cc_feature_origin)
ohMARRIAGE = encoder_marriage.fit(cc_feature_origin).transform(cc_feature_origin)
#ohSEX.take(5)

# 7.4 Pipeline

In [109]:
from pyspark.ml import Pipeline, PipelineModel
cc_numPipeline = Pipeline().setStages([cc_va,scaler_cc])
cc_catPipeline = Pipeline().setStages([encoder_sex, encoder_education, encoder_marriage])
cc_pipeline = Pipeline().setStages([cc_numPipeline, cc_catPipeline])
newcc = cc_pipeline.fit(cc_data).transform(cc_data)
#newcc.show(5)
cc_feature = VectorAssembler().setInputCols(["featured_scaled", "SEX_one-hot", "EDUCATION_one-hot", "MARRIAGE_one-hot"]).setOutputCol("cc_features")
cc_dataset = cc_feature.transform(newcc).select("cc_features", "DEFAULT")
cc_dataset.take(5)

[Row(cc_features=SparseVector(34, {0: 0.1541, 1: 4.0889, 2: 2.5305, 3: 1.9158, 4: 2.6036, 5: 1.7797, 6: 1.6706, 7: -0.8355, 8: -0.8553, 9: -1.7649, 10: -1.7391, 11: 0.0531, 12: 0.0436, 13: 0.0099, 18: 0.0299, 27: 1.0, 32: 1.0}), DEFAULT=1),
 Row(cc_features=SparseVector(34, {0: 0.9249, 1: 4.0889, 2: 2.5305, 3: 3.8316, 4: 2.8206, 5: -0.8898, 6: 1.6706, 10: 1.7391, 11: 0.0364, 12: 0.0242, 13: 0.0387, 14: 0.0509, 15: 0.0568, 16: 0.0548, 18: 0.0434, 19: 0.0568, 20: 0.0638, 22: 0.1125, 27: 1.0, 33: 1.0}), DEFAULT=1),
 Row(cc_features=SparseVector(34, {0: 0.6937, 1: 4.0889, 2: 2.5305, 3: 3.8316, 4: 3.6885, 11: 0.3971, 12: 0.1971, 13: 0.1955, 14: 0.2228, 15: 0.2459, 16: 0.2611, 17: 0.0916, 18: 0.0651, 19: 0.0568, 20: 0.0638, 21: 0.0655, 22: 0.2813, 27: 1.0, 33: 1.0}), DEFAULT=0),
 Row(cc_features=SparseVector(34, {0: 0.3854, 1: 4.0889, 2: 2.5305, 3: 1.9158, 4: 4.0139, 11: 0.6381, 12: 0.6777, 13: 0.7108, 14: 0.4401, 15: 0.4763, 16: 0.4961, 17: 0.1207, 18: 0.0876, 19: 0.0682, 20: 0.0702, 21: 0.

# 7.5 Make the models

In [110]:
[cc_trainSet, cc_testSet] = cc_dataset.randomSplit([0.8, 0.2])

## Logistic Regression

In [111]:
from pyspark.ml.classification import LogisticRegression

# train the model
cc_lr = LogisticRegression().setFeaturesCol("cc_features").setLabelCol("DEFAULT")
cc_lrModel = cc_lr.fit(cc_trainSet)
cc_lr_prediction = cc_lrModel.transform(cc_testSet)
cc_lr_prediction.show(5)

+--------------------+-------+--------------------+--------------------+----------+
|         cc_features|DEFAULT|       rawPrediction|         probability|prediction|
+--------------------+-------+--------------------+--------------------+----------+
|(34,[0,1,2,3,4,5,...|      0|[2.61263870995244...|[0.93167056938270...|       0.0|
|(34,[0,1,2,3,4,5,...|      0|[1.03776925665004...|[0.73841935337122...|       0.0|
|(34,[0,1,2,3,4,5,...|      0|[2.96992559691255...|[0.95119682321967...|       0.0|
|(34,[0,1,2,3,4,5,...|      0|[1.7008882108988,...|[0.84565070476470...|       0.0|
|(34,[0,1,2,3,4,5,...|      1|[-0.8654326762411...|[0.29620555497287...|       1.0|
+--------------------+-------+--------------------+--------------------+----------+
only showing top 5 rows



In [112]:
# cc_lr_total = cc_lr_prediction.count()
# cc_lr_correct = cc_lr_prediction.where("DEFAULT = prediction").count()
# cc_lr_wrong = cc_lr_prediction.where("DEFAULT != prediction").count()

# print(f'total: {cc_lr_total}, correct: {cc_lr_correct}, wrong: {cc_lr_wrong}')

In [113]:
from pyspark.ml.evaluation import BinaryClassificationEvaluator
# from pyspark.ml.evaluation import MulticlassClassificationEvaluator
cc_lr_prediction.select("prediction", "DEFAULT", "cc_features").show(5)
cc_evaluator = BinaryClassificationEvaluator().setRawPredictionCol("prediction").setLabelCol("DEFAULT")\
                                   .setMetricName("areaUnderROC")
auc = cc_evaluator.evaluate(cc_lr_prediction)
auc

+----------+-------+--------------------+
|prediction|DEFAULT|         cc_features|
+----------+-------+--------------------+
|       0.0|      0|(34,[0,1,2,3,4,5,...|
|       0.0|      0|(34,[0,1,2,3,4,5,...|
|       0.0|      0|(34,[0,1,2,3,4,5,...|
|       0.0|      0|(34,[0,1,2,3,4,5,...|
|       1.0|      1|(34,[0,1,2,3,4,5,...|
+----------+-------+--------------------+
only showing top 5 rows



0.6075852921954691

## Decision tree

In [114]:
from pyspark.ml.classification import DecisionTreeClassifier
cc_dt = DecisionTreeClassifier().setFeaturesCol("cc_features").setLabelCol("DEFAULT")
cc_dtModel = cc_dt.fit(cc_trainSet)

dt_predictions = cc_dtModel.transform(cc_testSet)
dt_predictions.show(5)
dt_evaluator = BinaryClassificationEvaluator().setRawPredictionCol("prediction").setLabelCol("DEFAULT")

auc = dt_evaluator.evaluate(dt_predictions)
auc

+--------------------+-------+----------------+--------------------+----------+
|         cc_features|DEFAULT|   rawPrediction|         probability|prediction|
+--------------------+-------+----------------+--------------------+----------+
|(34,[0,1,2,3,4,5,...|      0|[16756.0,2803.0]|[0.85669001482693...|       0.0|
|(34,[0,1,2,3,4,5,...|      0|[16756.0,2803.0]|[0.85669001482693...|       0.0|
|(34,[0,1,2,3,4,5,...|      0|[16756.0,2803.0]|[0.85669001482693...|       0.0|
|(34,[0,1,2,3,4,5,...|      0|[16756.0,2803.0]|[0.85669001482693...|       0.0|
|(34,[0,1,2,3,4,5,...|      1|   [221.0,751.0]|[0.22736625514403...|       1.0|
+--------------------+-------+----------------+--------------------+----------+
only showing top 5 rows



0.6524668221464371

## Random forest

In [118]:
from pyspark.ml.classification import RandomForestClassifier
cc_rf = RandomForestClassifier().setFeaturesCol("cc_features").setLabelCol("DEFAULT")
cc_rfModel = cc_rf.fit(cc_trainSet)

rf_predictions = cc_rfModel.transform(cc_testSet)
rf_predictions.show(5)
rf_evaluator = BinaryClassificationEvaluator().setRawPredictionCol("prediction").setLabelCol("DEFAULT")

auc = rf_evaluator.evaluate(rf_predictions)
auc

+--------------------+-------+--------------------+--------------------+----------+
|         cc_features|DEFAULT|       rawPrediction|         probability|prediction|
+--------------------+-------+--------------------+--------------------+----------+
|(34,[0,1,2,3,4,5,...|      0|[16.7782531185200...|[0.83891265592600...|       0.0|
|(34,[0,1,2,3,4,5,...|      0|[16.5892198102334...|[0.82946099051167...|       0.0|
|(34,[0,1,2,3,4,5,...|      0|[16.5039218856006...|[0.82519609428003...|       0.0|
|(34,[0,1,2,3,4,5,...|      0|[14.3248192329852...|[0.71624096164926...|       0.0|
|(34,[0,1,2,3,4,5,...|      1|[5.54542086608793...|[0.27727104330439...|       1.0|
+--------------------+-------+--------------------+--------------------+----------+
only showing top 5 rows



0.6327137600535219

## The AreaUnderROC for test sets:
We used three different models(logistic regression, decision tree, and random forest) to train a model to predict the target varible and compare the model's performances(AUC). Here are the AUC values for each model:

Logistic Regression: 0.6075852921954691

Decision Tree: 0.6524668221464371

Random Forest: 0.6327137600535219

Notice: The AUC value will change a bit beacuse of random trainset and testset. But after times of iterations, the random forest regression stiil behave best among these 3 models. 

So we choose Decision Tree model. 



# 7.6 Tuning

## We try to use the CrossValidator to select the best Random Forest model.We want to do the search among the different number of trees (5, 10, and 20), and different tree depth (5, 10, and 20).

In [116]:
from pyspark.ml.tuning import ParamGridBuilder, CrossValidator

paramGrid = ParamGridBuilder().addGrid(cc_rf.numTrees, [5,10,20]).addGrid(cc_rf.maxDepth,[5,10,20]).build()
evaluator = BinaryClassificationEvaluator().setLabelCol("DEFAULT").setRawPredictionCol("prediction")
cv = CrossValidator().setEstimator(cc_rf).setEstimatorParamMaps(paramGrid).setEvaluator(evaluator)

cc_cvModel = cv.fit(cc_trainSet)

predictions = cc_cvModel.transform(cc_testSet)
predictions.select("prediction", "DEFAULT", "cc_features").show(5)

auc = evaluator.evaluate(predictions)

print(auc)

+----------+-------+--------------------+
|prediction|DEFAULT|         cc_features|
+----------+-------+--------------------+
|       0.0|      0|(34,[0,1,2,3,4,5,...|
|       0.0|      0|(34,[0,1,2,3,4,5,...|
|       0.0|      0|(34,[0,1,2,3,4,5,...|
|       0.0|      0|(34,[0,1,2,3,4,5,...|
|       1.0|      1|(34,[0,1,2,3,4,5,...|
+----------+-------+--------------------+
only showing top 5 rows

0.6488201162235524


## As a result, the auc increases a bit from 0.6327137600535219 to 0.6488201162235524